chore: update getting started (#451)
knocklabs/docs: content/getting-started/what-is-knock.mdx

---

Knock is a hosted notifications engine that helps you implement notifications your users will love, without the effort of building and maintaining your own in-house notifications system.

In this overview, we’ll cover some of the foundational concepts of Knock. Knock is designed with both developers and product teams in mind: it’s easy for developers to implement quickly, and simple for less-technical users to maintain with our intuitive dashboard.

<div className="relative h-0" style={{ paddingBottom: "56.25%" }}>
  <iframe
    src="https://www.loom.com/embed/2865bc7d3bd04fcea9a527030c0f72b1"
    frameBorder="0"
    allowFullScreen
    className="absolute top-0 left-0 w-full h-full"
  ></iframe>
</div>
## Workflows

Workflows are a foundational concept in Knock. They allow you to easily model complex messaging flows across channels using a variety of logical function steps while respecting a user’s individual preferences. All Knock notifications are sent by triggering a workflow.

<Image
  src="/images/what-is-knock/workflow-diagram.png"
  alt="An image of a workflow diagram"
  width={1356}
  height={1452}
  className="rounded-md shadow-sm"
></Image>

> Your application can trigger workflows using our REST API, any one of our available SDKs (Node.js, .NET, PHP, Go, Elixir, .NET (C#), Java, or Python), or by integrating a CDP like Segment as an event source:

Author: Since I included the multilang code block, I think we can remove the enumeration. I link to sdks page and point out dropdown lang switcher

## Recipients

Recipients are in most cases users in your application. As you trigger workflows for recipients, Knock creates a cache of the data needed to notify them on different platforms, like an email address, phone number, avatar URL, or push token. Knock also stores custom properties you pass from your application to customize their notifications, like a plan type, user role, or timezone.

```javascript title="An object used to create a User"
{
	// Id is a required prop
	id: "1",
	// Knock also supports default props for common channels
	name: "John Hammond",
	email: "hammondj@ingen.net",
	phone_number: "555-555-5555",
	avatar: "https://ingen.net/headshots/hammondj.jpg",
	timezone: "America/Costa_Rica"
	// You can add as many custom props as needed. These will be
	// merged onto the top-level User object
	properties: {
		"title": "CEO",
		"planType": "allAccess",
		"userType": "admin"
	}
}
```

## Channels

Channels in Knock represent a specific provider you have configured to send notifications. You can include channel steps in your workflows to send notifications with the providers you already use in production.
Knock supports the following channel types and providers:

> <AccordionGroup>

Author: Good callout.

    Native)](/integrations/push/expo), and [Firebase Cloud Messaging
    (Android)](/integrations/push/firebase).
  </Accordion>
  <Accordion title="Chat">
    Knock supports sending chat messages with [Slack](/integrations/chat/slack),
    [Discord](/integrations/chat/discord), [Microsoft
    Teams](/integrations/chat/microsoft-teams),
    [WhatsApp](/integrations/chat/whatsapp).
  </Accordion>
  <Accordion title="In-app">
    Knock provides [a real-time in-app feed API](/integrations/in-app/knock) for
    receiving notifications, along with drop-in components to display them to
    your users.
  </Accordion>
</AccordionGroup>

When a channel step generates a notification as a part of a workflow that’s called a `message`, and Knock allows you to define dynamic message templates using a combination of a drag-and-drop editor and the liquid templating language.

This helps product and marketing teams standardize on one templating system instead of using different templating languages for different providers. It also has the added benefit of lifting these messages out of your codebase so you can iterate quickly on customer communications without a developer.

> ## Functions

Author: I think I have a different take here, mainly that we'd want to discourage side quests where possible. If we link off, we may not get them back in this flow.

> ## Functions

Reviewer: I agree with Jeff for the most part here. Can we link out to the overview page of designing workflows at the end of this block though? That'll let someone new-tab the overview but continue here until later.

> ## Functions

Reviewer 2: > If we link off, we may not get them back in this flow.

> ## Functions

Reviewer 2: > If we link off, we may not get them back in this flow.
This is interesting, I guess I usually take the approach of giving the user as much choice as possible to make it easer/more-intuitive for them, but I'm certainly no UX expert. My personal preference when navigating a site is to be able to "choose your own adventure" to whatever I'm interested in, so that if something piques my interest I can find out more information.

For me this boils down to the root question I raised in another place which is "how much information do we want on this page." IMO, if we give partial information about something (like a general description of a more-complex feature), we should make it easy for a user to find out more about it.

> ## Functions

Reviewer 2: I guess my main concern here is, if we do something like say "you can use a batch function!" and that's exactly what a potential customer needs but we don't make it easy for them to find out exactly how the batch function works, that might not be a great experience for them

> ## Functions

Author: With a bunch of content in the accordions, I went back and included links to most of the concepts. I think this does a good job of splitting the difference between enough info that someone could read just this and know 80% of what they need to about Knock, and easily digging into more details if they need to.

Each workflow can combine multiple function steps to model complex logic that creates better notification experiences.

> You can combine the following function steps with any number of channel steps to create great notifications:

Author: This is a good callout. I think my take here is that we need some level of detail so that using just this page, someone can tell if Knock can do what they want. But simplification is the name of the game, so I leaned into using the accordion here. That way it's still very scannable, and users can opt into the weeds that seem important to them. Interested to here your thoughts.

**Examples:**

- Only send an email message if an in-app message has not been seen
- Only send an in-app notification if `recipient.plan === "pro"`
- Only execute a delay step if `delay === true` in the trigger payload

In addition to giving your technical and non-technical users the ability to construct these workflows via a drag-and-drop editor, Knock also enables your users to exercise control over their own notification experience using a flexible preferences model.

## Preferences

In Knock, each workflow run is executed on behalf of a recipient, and each recipient can specify their preferences to receive notifications across a number of different criteria: channel types, individual workflows, and workflow categories.

<Image
  src="/images/what-is-knock/preferences.png"
  alt="An image of a preference set"
  width={1356}
  height={1452}
  className="rounded-md"
></Image>

Application developers have control over how these preference sets are presented to the user and which options to surface, but Knock enforces these preferences during every workflow run automatically.

> Let’s take a look at some examples of the different preference types:

Author: Did the same with the accordion here, opt into complex explanation

While the workflow engine is at the heart of Knock, our goal is to build a complete notification system for our users. Here is an overview of some of the more advanced features that we provide:

#### UI components

Knock provides developers with React components like `<NotificationFeed/>` and `<SlackAuthButton/>` to use in their applications. You can read more about [building in-app UI](/in-app-ui/overview) for the web and mobile with Knock to get started.

#### Advanced concepts

There is a lot more to learn about Knock, and our [concepts overview page](/concepts/overview) is a good place to start. Here are use cases our customers commonly solve with Knock:

- Powering [translation and localization](/concepts/translations) and managing timezone aware delivery
- Creating advanced notification logic using [subscriptions](/concepts/subscriptions) and [schedules](/concepts/schedules)
- Joining Knock with your application's data model using [tenants](/concepts/tenants) and [objects](/concepts/objects) to power customized experiences
- Using [the template editor](/designing-workflows/template-editor/overview) to standarize messaging templates across providers

#### Developer tools

> Knock is a developer-first platform, with both [environment](/concepts/environments) and [commit models](/concepts/commits). If you want to work with Knock resources in code, you can use our [Management API](/developer-tools/management-api) or [CLI](/developer-tools/knock-cli). Once you are sending notifications through Knock, you can also [use workflow run logs](/send-notifications/debugging-workflows) to examine all steps of workflow execution or stream data into a monitoring system like Datadog with [extensions](/integrations/extensions/overview).

Reviewer 2: ```suggestion
Knock is a developer-first platform, with both [environment](/concepts/environments) and [commit models](/concepts/commits). If you want to work with Knock resources in code, you can use our [Management API](/developer-tools/management-api) or [CLI](/developer-tools/knock-cli). 

Once you're sending notifications through Knock, we offer observability tools like [workflow run logs](/send-notifications/debugging-workflows) (to examine all steps of workflow execution in your dashboard) and data streaming into a monitoring system like Datadog with [extensions](/integrations/extensions/overview).
```

---
title: What is Knock?
description: Learn more about what Knock does and how it helps power your product notifications.
tags: ["getting started", "explainer", "explained"]
section: Getting started
---

> Knock is a hosted notifications engine that helps you implement notifications your users will love, without the effort of building and maintaining your own in-house notifications system.

Reviewer: Let's be consistent here and use notifications infrastructure as what Knock is.

  <Accordion title="Branch">
    [A branch step](/designing-workflows/branch-function) uses multiple
    conditions to execute different branches of logic, e.g. if `user.planType
    === 'pro'` send them email A, else send them email B.
  </Accordion>
  <Accordion title="Throttle">
    [A throttle step](/designing-workflows/throttle-function) controls how many
    times a user is notified for a particular workflow over a specified
    duration, e.g. trigger the `server is down` workflow every minute while the
    server is down, but only send a max of one email every 5 minutes.
  </Accordion>
  <Accordion title="Fetch">
    [A fetch step](/designing-workflows/fetch-function) makes an HTTP request to
    an external service and uses the returned data in subsequent steps, e.g.
    query an MLS API for recent home sales in the user’s zip code and render
    them in an email.
  </Accordion>
</AccordionGroup>

In addition to combining channel steps and function steps to create complex workflows, you can augment these steps with additional logic based on the user recipient, inputs from your application, or the status of previous workflow steps. These are called step conditions.

> ## Step Conditions

Reviewer 3: ```suggestion
## Step conditions
```
Title case --> sentence case

  </Accordion>
</AccordionGroup>

> When a channel step generates a notification as a part of a workflow that’s called a `Message`, and Knock allows you to define dynamic message templates using a combination of a drag-and-drop editor and the Liquid templating language.

Reviewer 2: I know I'm being a bit pedantic (😅 ) but I think this sentence structure could still be a bit better grammar-wise:

```suggestion
A notification that is generated by a channel step as a part of a workflow is called a `Message`, and Knock allows you to define dynamic message templates using a combination of a drag-and-drop editor and the Liquid templating language.
```

Author: applied in another commit

Reviewer 2: @JEverhart383 I think the wording is different? The "that's" is what was bugging me here

Author: @cellomatt fail on my end, didn't save that file. Should be good now
